Add Cisco model-driven telemetry & GNMI inputs

sbyx opened the pull request:
This adds input plugins for Cisco model-driven telemetry & GNMI-based telemetry.

### Required for all PRs:

- [ ] Signed [CLA](https://influxdata.com/community/cla/).
- [ ] Associated README.md updated.
- [ ] Has appropriate unit tests.

glinton commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,504 @@
+
+	// Decode Telemetry field value if set
+	var value interface{}
+	switch field.ValueByType.(type) {
```
Similar to gnmi, assign a value in the switch statement and use in the case.

sbyx commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,504 @@
+
+	// GRPC TLS settings
+	TLS bool
+	internaltls.ServerConfig
```
It is diffierentiated by the transport type. dialin transport uses ClientConfig and dialout transports use ServerConfig (depending on if telegraf acts as a client or server for telemetry).

glinton replied:
Right, but the toml library that parses it out of the config will fail as it won't know which struct to assign the values to.

You could add all the settings into `CiscoTelemetryMDT` then where you need it:
```go
type CiscoTelemetryMDT struct {
	TLSCA              string   `toml:"tls_ca"`
	TLSCert            string   `toml:"tls_cert"`
	TLSKey             string   `toml:"tls_key"`
	InsecureSkipVerify bool     `toml:"insecure_skip_verify"`
	TLSAllowedCACerts  []string `toml:"tls_allowed_cacerts"`
...
```

```go
tlsConfig, err := internaltls.ServerConfig{
	TLSCert: c.TLSCert
	TLSKey: c.TLSKey
	TLSAllowedCACerts: c.TLSAllowedCACerts
}.TLSConfig()
```

sbyx replied:
Understood, I remember doing some tests in dialout aka. server mode with TLS at some point, but I am not sure if my collegaues tested TLS with client mode so it may have slipped through. I changed the code according to your suggestions. See new commits below.

sbyx commented:
Thanks a lot for your feedback. I tried to incorporate most of you comments with some new commits.

glinton commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,504 @@
+	var mutex sync.Mutex
+	clients := make(map[net.Conn]struct{})
+
+	for {
```
Can you check the other instances where similar `for {` loops are defined, it seems like the context is only checked on an error within the loop.

sbyx replied:
Done, thanks!

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,411 @@
+					}
+				}
+
+				tags["Producer"] = c.ServiceAddress
```
No big preference on my side, please let me know what you prefer and I will change it.

sbyx replied:
Fine by me.

glinton commented:
Thanks, could you look into the panic in the tests:
```
2019/05/16 06:57:11 I! Started Cisco MDT service on 127.0.0.1:57002
panic: runtime error: invalid memory address or nil pointer dereference
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x82628d]

goroutine 131 [running]:
github.com/influxdata/telegraf/vendor/google.golang.org/grpc.(*listenSocket).Close(0xc0003720e0, 0xc0003ea1e0, 0xc000290568)
	/go/src/github.com/influxdata/telegraf/vendor/google.golang.org/grpc/server.go:489 +0x2d
github.com/influxdata/telegraf/vendor/google.golang.org/grpc.(*Server).Serve.func2(0xc000438000, 0xc0003720e0)
	/go/src/github.com/influxdata/telegraf/vendor/google.golang.org/grpc/server.go:535 +0xb8
panic(0x8eb8e0, 0xda36c0)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/influxdata/telegraf/vendor/google.golang.org/grpc.(*Server).Serve(0xc000438000, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/influxdata/telegraf/vendor/google.golang.org/grpc/server.go:544 +0x1fe
created by github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt.TestGRPCDialinError
	/go/src/github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:448 +0x11f
FAIL	github.com/influxdata/telegraf/plugins/inputs/cisco_telemetry_mdt	5.188s
```

sbyx commented:
I have rewritten the synchronization aspects between server and client in the tests. I believe there were some race conditions which are solved now and overall tests run much faster now. This also solved the panic issue you mentioned.

glinton commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -164,6 +164,12 @@ func (c *CiscoTelemetryGNMI) subscribeGNMI(client *grpc.ClientConn) {
 
+				// Unable to derive the individual measurements or subscription without a prefix
+				if response.Update.Prefix == nil {
+					c.acc.AddError(fmt.Errorf("GNMI update with missing prefix discarded"))
```
To me, this would be better as a printed debug log or removed completely.

sbyx replied:
Removed the error message.

danielnelson commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,516 @@
+	}
+
+	for c.ctx.Err() == nil {
+		packet, err := stream.Recv()
```
I'm not super familiar with gRPC, but how does this call to Recv get cancelled on shutdown?  Is it interrupted by the context or does it require reading a message?

I'm concerned this will hang if no further messages are received on the stream.

sbyx replied:
You may be right, I have switched to using grpc.Server Stop() which should take care of that:

> Stop stops the gRPC server. It immediately closes all open connections and listeners. It cancels all active RPCs on the server side and the corresponding pending RPCs on the client side will get notified by connection errors.

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+				tags["Producer"] = c.ServiceAddress
+				tags["Target"] = response.Update.Prefix.Target
+				builder.Truncate(builder.Len() - 1)
+				prefix := builder.String()
```
The problem is that especially for multiple subscriptions per plugin we will receive measurements with different data structures on the same stream and there is not really a good way to differentiate other than the Prefix returned by the device.

If we were to aggregate all incoming data on the stream under the same measurement name subsequent measurements might have totally different key- or value-fields which I guess especially for InfluxDB would create a big mess.

Therefore the idea was to use the (XPath-like) path-name here to differentiate measurements since it is sort of a natural identifier for this kind of telemetry data.

Same applies more or less to the MDT plugin but for that in general it is a bit more contained environment compared to GNMI where you have different vendors on top of the existing ambiguities of the protocol.

danielnelson replied:
I'd really like to get the data into a format that is more typical to InfluxDB to it can coexist with other Telegraf metrics nicely.  To me this means having a limited set of measurement names, most plugins have 1-5, and being able to lookup the metrics and group using tags.  The measurement name should be a single word namespace for the metrics.

We may need to look over several examples.  Taking just the ones from the README I would recommend just moving the path to a tag.
```diff
- openconfig:/interfaces/interface/state/counters,name=GigabitEthernet0/0/0/0,source=10.49.234.115
+ openconfig,path=/interfaces/interface/state/counters,name=GigabitEthernet0/0/0/0,source=10.49.234.115
```

It probably gets more interesting with keys in the path though, can you show some more complex examples?

sbyx commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,516 @@
+	tcpMaxMsgLen uint32 = 1024 * 1024
+
+	// IOS XR EMS dialin telemetry GPBKV encoding
+	grpcEncodeGPBKV int64 = 3
```
It is the encoding requested by the device. Some devices support multiple encodings and some actually require individual .proto files per data model stream. This plugin supports the more or less universal encoding GPBKV according to the "message TelemetryField" in the telemetry/telemetry.proto message.

sbyx commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,516 @@
+func (c *CiscoTelemetryMDT) MdtDialout(stream dialout.GRPCMdtDialout_MdtDialoutServer) error {
+	peer, peerOK := peer.FromContext(stream.Context())
+	if peerOK {
+		log.Printf("D! Accepted Cisco MDT GRPC dialout connection from %s", peer.Addr)
```
We are only using this info for debug log, so it is not essential. Also this should not fail, check is just a precaution ;)

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+/**
+ * Copyright (c) 2018 Cisco Systems
+ * Author: Steven Barth <stbarth@cisco.com>
+ */
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+	Redial internal.Duration
+
+	// GRPC TLS settings
+	TLS bool
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+	var err error
+	var opts []grpc.DialOption
+	c.acc = acc
+	c.ctx, c.cancel = context.WithCancel(context.Background())
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+
+	client, err := grpc.Dial(c.ServiceAddress, opts...)
+	if err != nil {
+		return fmt.Errorf("E! Failed to dial GNMI: %v", err)
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+	}
+
+	client.Close()
+	c.wg.Done()
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+
+	// Dialin client telemetry stream reading routine
+	c.wg.Add(1)
+	go c.subscribeGNMI(client)
```
Yeah I slightly adjusted it since it makes more sense with a loop, otherwise we would have a potentially prematurely terminating plugin again.

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+}
+
+// SubscribeGNMI and extract telemetry data
+func (c *CiscoTelemetryGNMI) subscribeGNMI(client *grpc.ClientConn) {
```
Done, I split it up at the main for-loops.

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+			log.Printf("D! Connection to GNMI device %s closed", c.ServiceAddress)
+		}
+
+		if c.Redial.Duration.Nanoseconds() <= 0 {
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+	c.cancel()
+	c.wg.Wait()
+
+	log.Println("I! Stopped GNMI service on ", c.ServiceAddress)
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+	Encoding    string
+	Origin      string
+	Prefix      string
+	Target      string
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/README.md:
```diff
@@ -0,0 +1,47 @@
+
+  [[inputs.cisco_telemetry_gnmi.subscription]]
+    origin = "Cisco-IOS-XR-infra-statsd-oper"
+    path = "infra-statistics/interfaces/interface/latest/generic-counters"
```
Done hopefully, this is a bit platform specific but I added some hints to the relevant GNMI documentation.

sbyx replied:
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+					UpdatesOnly:  c.UpdatesOnly,
+				},
+			},
+		}
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_mdt/README.md:
```diff
@@ -0,0 +1,51 @@
+
+The GRPC dialin transport is supported on IOS XR (64-bit) 6.1.x and later.
+
+The TCP dialout transport is supported on IOS XR (32-bit and 64-bit) 6.1.x and later.
```
I removed the dial-in for now, since it is of lesser importance for now. I will talk to my colleagues and see if that is fine for now or if we add this again add a later stage.

As for the other comments I tried to apply similar changes I have made to GNMI based on your comments to this plugin as well.

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+
+// CiscoTelemetryGNMI plugin instance
+type CiscoTelemetryGNMI struct {
+	ServiceAddress string         `toml:"service_address"`
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_mdt/ems/ems.pb.go:
```diff
@@ -0,0 +1,1742 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
```
I removed ems since it belonged to dialin (see above).
There is unfortunately no good updated repository here for this purpose. Some carry the latest .proto files but no up-to-date .pb.go.

danielnelson replied:
Ideally everything in this repository would be only under the top MIT license, while we can still distribute this code as Apache 2.0 it does cause some confusion and difficulties keeping track of more than one license in the repository.  Could you either license these as MIT (essentially removing this header) or move these pb.go and .proto files to a new github repository?

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+}
+
+//ParsePath from XPath-like string to GNMI path structure
+func parsePath(origin string, path string, target string) *gnmi.Path {
```
Made some modifications there

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,416 @@
+					}
+
+					if value != nil {
+						fields[builder.String()] = value
```
Please see latest additions to README.md

danielnelson commented on plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:
```diff
@@ -0,0 +1,380 @@
+		}
+
+		if hdr.MsgLen > maxMsgSize {
+			err = fmt.Errorf("dialout packet too long: %v", hdr.MsgLen)
```
`return fmt.Errorf(...`, do this at other break points in this function.

sbyx replied:
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+		}
+		subscriptions[i] = &gnmi.Subscription{
+			Path:              gnmiPath,
+			Mode:              gnmi.SubscriptionMode(gnmi.SubscriptionMode_value[strings.ToUpper(subscription.SubscriptionMode)]),
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+			Subscribe: &gnmi.SubscriptionList{
+				Prefix:       gnmiPath,
+				Mode:         gnmi.SubscriptionList_STREAM,
+				Encoding:     gnmi.Encoding(gnmi.Encoding_value[strings.ToUpper(c.Encoding)]),
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+
+	err = subscribeClient.Send(request)
+	if err != nil {
+		return fmt.Errorf("GNMI subscription setup failed: %v", err)
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+		reply, err = subscribeClient.Recv()
+		if err != nil {
+			if err != io.EOF && ctx.Err() == nil {
+				err = fmt.Errorf("GNMI subscription aborted: %v", err)
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+			if err != io.EOF && ctx.Err() == nil {
+				err = fmt.Errorf("GNMI subscription aborted: %v", err)
+			} else {
+				err = nil
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+	for i := 0; i < len(path); i++ {
+		if path[i] == '[' {
+			if name >= 0 {
+				fmt.Println(path, i, name, start, end)
```
Ooops, sorry about that.

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,456 @@
+		return fmt.Errorf("GNMI subscription setup failed: %v", err)
+	}
+
+	log.Printf("D! Connection to GNMI device %s established", c.Address)
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_mdt/README.md:
```diff
@@ -0,0 +1,38 @@
+  service_address = ":57000"
+
+  ## Enable TLS for GRPC transport
+  # enable_tls = true
```
Done

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/README.md:
```diff
@@ -0,0 +1,67 @@
+  # tls_key = "/etc/telegraf/key.pem"
+
+  ## GNMI encoding requested (usually one of: "proto", "json", "json_ietf")
+  # encoding = "proto"
```
I have excluded those types.

sbyx commented:
Will follow-up tomorrow with the remaining points and check test regressions

sbyx commented on plugins/inputs/cisco_telemetry_gnmi/README.md:
````diff
@@ -0,0 +1,61 @@
+```toml
+[[inputs.cisco_telemetry_gnmi]]
+  ## Address and port of the GNMI GRPC server
+  address = "10.49.234.114:57777"
````
You can already add multiple instances of the plugin pointing to different devices.

danielnelson replied:
While multiple plugins can be used, it could be nice from a configuration point of view to be able to list multiple devices to reduce the duplication.

One downside of connecting to multiple devices in a single plugin instance is that if one device blocks the input then it can prevent other devices from reporting, though this is mostly an issue with polling plugins.  For this plugins in particular the main downside would be more complexity managing multiple connections.

We should consider how many devices one may want to monitor, especially with the same set of subscriptions.

sbyx replied:
Okay, I added it in https://github.com/influxdata/telegraf/pull/5852/commits/eb13faa4a7ad8f55b51712c7d556aae5f55ae7bb it wasn't that complex.

danielnelson commented:
Regarding how metrics should be created via `AddFields()`, here is the transformations I think we should do:

1. For each subscription the user should be able to provide a measurement name, if not specified the default name will be the last path segment of the subscription path.  The configuration could look like this:

   ```
   [[inputs.cisco_telemetry_gnmi]]
     [[inputs.cisco_telemetry_gnmi.subscription]]
   ￼   path = "/acl"
       name = "acl"
   ```

2. The subscription path should be added as a tag called path.

   ```
   acl,path=/acl
   ```

3. For each field, the relative path from the subscription path is used.

   So if the full path is:
   ```
   /interfaces/interface/subinterfaces/subinterface/state/counters/in-pkts
   ```

   and the subscription is:
   ```
   /interfaces/interface/
   ```

   the field key would contain only the remaining relative path:
   ```
   subinterfaces/subinterface/state/counters/in_pkts
   ```

4. All keys in the paths are extracted as tags and not included in the field key or measurement name.  I have the understanding that a path would never reuse key names so it there would never be a tag key conflict.  For example this would never be happen:
   ```
   /interface[name=foo]/subinterface[name=bar]
   ```

5. Replace `-` in fields with `_`.  The `-` is a special character as well and restricted in Prometheus format, in InfluxDB queries it must be quoted to avoid being interpreted as the subtraction operator.  I feel like this is a rather safe transformation.

   Quite a number of the outputs do not support the `/` character, such as Prometheus, OpenTSDB, Graphite.  Additionally in InfluxDB it has special meaning and must be quoted in queries.  However, it's not clear if there is a good conversion to make for this character, so I think we leave it as `/` and encourage outputs to make an appropriate conversion for them.  Alternately a user could use the strings processor to for a particular form.


Here is more complete example of all transformations:

```toml
[[inputs.cisco_telemetry_gnmi]]
  [[inputs.cisco_telemetry_gnmi.subscription]]
    path = "/acl"
    name = "acl"
```

Example path:
```
/acl/acl-sets/acl-set[name=VLAN13_OUT][type=ACL_IPV4]/acl-entries/acl-entry[sequence-id=4090]/state/matched-packets
```

Example metric in InfluxDB line protocol:
```
acl,path=/acl,name=VLAN13_OUT,type=ACL_IPV4,sequence_id=4090 acl_sets/acl_set/acl_entries/acl_entry/state/matched_packets=0i
```

@sbyx @littlespace Let me know if you see issues with this transformation.

sbyx commented:
@danielnelson I tried to incorporate all your suggestions and also adapted them to the MDT plugin.
For GNMI I rewrote the prefix handling to be less strict with whatever GNMI implementations are out there. Let's go through your points quickly:

1. done, I added a generic aliasing table for the MDT side (since we are not actively subscribing here and just receiving), I also made it available to GNMI in addition to what you suggested, since it could be useful for more complex path combinations and platforms need not actually exactly return what requested but could go e.g. a level above or below.

2. done

3. done

4. It is unlikely to happen and unless it actually happens, we will use short-names. This is true for both GNMI and MDT.

5. Done for '-' to '_'

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,532 @@
+				break
+			}
+		}
+		if (tagsChanged || currentAliasPath != lastAliasPath) && len(fields) > 0 {
```
I think here you will want to use [metric.SeriesGrouper](https://github.com/influxdata/telegraf/blob/master/metric/series_grouper.go#L24).  Just keep adding tags and fields to this type and then at the end of this function you can range over grouper.Metrics() and call acc.AddMetric().

If I am understanding correctly this will allow you to skip the code to detect tag changes.

sbyx replied:
Done, see https://github.com/influxdata/telegraf/pull/5852/commits/2351ca080ebb02fcbb73cb329fcde10a1de41e73

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,532 @@
+}
+
+// Parse path to path-buffer and tag-field
+func (c *CiscoTelemetryGNMI) handlePath(path *gnmi.Path, tags map[string]string, builder *bytes.Buffer) string {
```
The life of the builder is very difficult for me to follow, as it moves into several functions and is truncated.  Instead of passing a bytes.Buffer in, I suggest creating a new bytes.Buffer in this function and returning the results of .String().  You could pass the prefix in to handle the individual updates.

The function would return the full path and the aliased path and add any path tags.

sbyx replied:
Got rid of passing around of the buffer.

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,532 @@
+}
+
+// HandleTelemetryField and add it to a measurement
+func (c *CiscoTelemetryGNMI) handleTelemetryField(update *gnmi.Update, tags map[string]string, builder *bytes.Buffer) (string, map[string]interface{}) {
```
Pass in a prefix path and create bytes.Buffer internally, returning the .String()

sbyx replied:
Got rid of passing around of the buffer.

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi_test.go:
```diff
@@ -0,0 +1,239 @@
+	assert.Nil(t, c.Start(acc))
+
+	server.Serve(listener)
+	c.Stop()
```
I think this will be racy, in case the data is not received before Stop is called.  You can use the acc.Wait(n) function to wait for a known number of metrics.

sbyx replied:
Done, thanks!

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi_test.go:
```diff
@@ -0,0 +1,239 @@
+	acc := &testutil.Accumulator{}
+	assert.Nil(t, c.Start(acc))
+	server.Serve(listener)
+	c.Stop()
```
To avoid test race, call acc.WaitError(n) before calling Stop

sbyx replied:
Done, thanks!

danielnelson commented on plugins/inputs/cisco_telemetry_gnmi/cisco_telemetry_gnmi.go:
```diff
@@ -0,0 +1,532 @@
+			defer c.wg.Done()
+			for ctx.Err() == nil {
+				if err := c.subscribeGNMI(ctx, address, tlscfg, request); err != nil {
+					acc.AddError(err)
```
I noticed that I can get an error when shutting down Telegraf (since I don't have a real device to connect to):
```
E! [inputs.cisco_telemetry_gnmi]: Error in plugin: failed to setup subscription: rpc error: code = Canceled desc = context canceled
```

Also, can you make sure to double check for errors and for successful reconnecting if we lose connection to the monitored device.

sbyx replied:
Suppressing errors now when the context is cancelled. As for reconnecting we are running the same steps as for the first connection over and over again (with the redial timeout inbetween) until the connection succeeds.

danielnelson replied:
:+1: Thanks, I see/remember the reconnect logic now

sbyx commented:
Okay, all done. There is not really a simulator only some virtualized routers I'm using for testing aside physical devices. I will run some more device-based tests later.
